Enable lpdac for lpcxpresso55s36

Albort12138 opened the pull request:
Created mcux lpdac driver;
Enabled dac example on lpcxpresso55s36;
Verify dac test suit on lpcxpresso55s36.

Albort12138 committed:
drivers: dac: Add driver for mcux lpdac

Create dac_mcxu_lpdac.c file to implement mcux lpdac, add binding for
the mcux lpdac, update Kconfig.mcux and CMakeLists.txt file to support
mcux lpdac.

Signed-off-by: Albort Xue <yao.xue@nxp.com>

Albort12138 commented:
Block by approve of https://github.com/zephyrproject-rtos/hal_nxp/pull/265

danieldegrasse commented on dts/arm/nxp/nxp_lpc55S3x_common.dtsi:
```diff
@@ -294,6 +294,33 @@
+		#io-channel-cells = <1>;
+	};
+
+	dac2: dac@B9000 {
```
nit- please use lowercase hex styling to align with the rest of this file

danieldegrasse commented on soc/arm/nxp_lpc/lpc55xxx/soc.c:
```diff
@@ -344,6 +344,26 @@ DT_FOREACH_STATUS_OKAY(nxp_lpc_ctimer, CTIMER_CLOCK_SETUP)
 #endif
 #endif
+#if DT_NODE_HAS_COMPAT_STATUS(DT_NODELABEL(dac0), nxp_lpdac, okay)
+#if defined(CONFIG_SOC_LPC55S36)
```
I'd prefer we don't duplicate the init code for the VREF here- instead, combine the conditional for the LPDAC and LPADC to set up the VREF if either peripheral is enabled

danieldegrasse commented on dts/bindings/dac/nxp, lpdac.yaml:
```diff
@@ -0,0 +1,29 @@
+  voltage-reference:
+    type: int
+    required: true
+    description: DAC voltage reference select
```
I believe that the meaning of this value is SOC specific- could we add something to the description here clarifying that, just to avoid ambiguity?

Albort12138 commented:
> @Albort12138 thank you for the PR, can you help address the other CI failures.

mmahadevan108 commented:
@Albort12138 Reopening the PR, I am assuming it was accidentally closed.

mmahadevan108 commented:
HAL side has merged, please update `west.yml`

mmahadevan108 commented on boards/arm/lpcxpresso55s36/lpcxpresso55s36-pinctrl.dtsi:
```diff
@@ -2,7 +2,7 @@
  * from LPC55S36.mex
  *
- * Copyright 2022 NXP
+ * Copyright 2022-2023 NXP
```
Please update the board documentation file to indicate support for this feature.

mmahadevan108 commented:
@Albort12138 , please take a look at the CI failures.

martinjaeger commented on drivers/dac/dac_mcux_lpdac.c:
```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright 2023, NXP
+ * All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
```
Why is this BSD licensed? If possible, please use Apache 2.0 and the same header style as everywhere else.

martinjaeger commented on boards/arm/lpcxpresso55s36/doc/index.rst:
```diff
@@ -81,7 +81,8 @@ already supported, which can also be re-used on this lpcxpresso55s36 board:
 +-----------+------------+-------------------------------------+
-
+| DAC       | on-chip    | dac                                 |
++-----------+------------+-------------------------------------+
```
Please keep newline after the table.

martinjaeger commented on boards/arm/lpcxpresso55s36/lpcxpresso55s36.dts:
```diff
@@ -175,3 +175,8 @@ zephyr_udc0: &usbfs {
 &dma0 {
 	status = "okay";
 };
+&dac0 {
```
Please add newline before `&dac0 {`.

martinjaeger commented on drivers/dac/dac_mcux_lpdac.c:
This file is almost exactly the same as `dac_mcux_dac.c` implemented by @henrikbrixandersen.

1. His copyright should be at to the top of this file.
2. Can we not re-use the same implementation to avoid code duplication? I don't know the NXP MCUs very well, but I see that the "normal" dac also has a `low_power` property. What exactly is the difference between `dac` and `lpdac`?

martinjaeger commented on samples/drivers/dac/README.rst:
```diff
@@ -209,6 +209,18 @@ also can run for the
+   :goals: build flash
+   :compact:
+
+DAC output is available on connector J12 pin 4.
```
Add newline after this line.

martinjaeger commented on dts/bindings/dac/nxp, lpdac.yaml:
There is a space in the file name.

martinjaeger commented on boards/arm/lpcxpresso55s36/doc/index.rst:
Please add the `dac` also to the `supported` features in `lpcxpresso55s36.yaml`. Otherwise, the tests are not run.

martinjaeger commented on tests/drivers/dac/dac_loopback/boards/lpcxpresso55s36.overlay:
```diff
@@ -0,0 +1,9 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+&dac0 {
```
Please add newline here.

martinjaeger commented on samples/drivers/dac/boards/lpcxpresso55s36.overlay:
```diff
@@ -0,0 +1,12 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+/ {
```
Please add newline here.

martinjaeger commented on tests/drivers/dac/dac_loopback/boards/lpcxpresso55s36.overlay:
The board must also be added to the `testcase.yml`.

martinjaeger commented on drivers/dac/dac_mcux_lpdac.c:
```diff
@@ -0,0 +1,107 @@
+
+struct mcux_lpdac_config {
+	LPDAC_Type *base;
+	dac_reference_voltage_source_t refVoltage;
```
Please use snake_case for variable names:

```suggestion
	dac_reference_voltage_source_t ref_voltage;
```

martinjaeger commented:
As mentioned before, I'm not familiar with the NXP MCU portfolio and I'm getting a bit confused by the naming of the MCU series and driver IPs.

Previous driver implementation uses the `mcux` prefix for the `.c` implementation, but `kinetis` in the bindings/compatibles (e.g. `nxp,kinetis-dac`). Are all NXP MCUs named MCUX and is `kinetis` the series? If yes, this driver should probably be prefixed with `lpc`, as LPC is the series of this one, right?

And where does the name `lpdac` come from? In the [LPC553x Reference Manual](https://cache.nxp.com/docs/en/reference-manual/LPC553xRM.pdf) it is just named `DAC`, as far as I can see.

@Albort12138 could you or someone else from NXP please clarify which different DAC IPs NXP has and to which MCU series they belong? This should also be mentioned in the binding comment.

decsny commented:
> Previous driver implementation uses the `mcux` prefix for the `.c` implementation, but `kinetis` in the bindings/compatibles (e.g. `nxp,kinetis-dac`). Are all NXP MCUs named MCUX and is `kinetis` the series? If yes, this driver should probably be prefixed with `lpc`, as LPC is the series of this one, right?

LPDAC is on more than just LPC, same as the one called kinetis is on more than just kinetis. Whoever implemented the "kinetis" one originally just called it that because they were probably doing it for a kinetis part. We are now trying to avoid using SOC series names in binding names for this reason, because the IP blocks are not specific to certain series, over time, they get re-used. If anything, the name of the kinetis one should be changed (there are a lot of nxp bindings in the tree that are not ideally named but for backward compatibility we don't change them willy-nilly)

> And where does the name lpdac come from? In the [LPC553x Reference Manual](https://cache.nxp.com/docs/en/reference-manual/LPC553xRM.pdf) it is just named DAC, as far as I can see.

The chapter is called, DAC, section 47.2 says the full name (Low Power Digital-to-Analog Converter), and introduces an abbreviation DAC to conventiently refer to within the RM and chapter (since there is only one type of DAC in the SOC, it's not confusing within that context). The NXP HAL calls it LPDAC in the device files from this full name to distinguish it from other dacs. I believe the driver folder cmake module however is called dac_1.

martinjaeger commented:
Ok, thanks @decsny for the explanation. Final question: What does `MCUX` stand for? Does this cover all NXP MCUs?

@Albort12138 can you please add a comment to the binding stating the MCU series this compatible covers (so far)? This will be helpful for people adding support for future MCUs and for people like me who are not so familiar with all these names and IPs. Also please explain what the abbreviation lpdac stands for (see above).

martinjaeger commented on drivers/dac/Kconfig.mcux:
```diff
@@ -19,6 +19,13 @@ config DAC_MCUX_DAC32
 	  Enable the driver for the NXP Kinetis MCUX DAC32.
 
+config DAC_MCUX_LPDAC
+	bool "NXP Kinetis MCUX LPDAC driver"
```
Given the discussion about naming we had before, this is probably wrong?

decsny replied:
I agree that we should remove "kinetis" at least from any terminology. "mcux" I think is loosely used too but comes from the name of the NXP MCU SDK which is called MCUX SDK.

martinjaeger replied:
Would `CONFIG_DAC_NXP_LPDAC` not be the most self-explanatory then? MCUX seems to be used also [by other vendors](https://www.businesswire.com/news/home/20230627745063/en/Bluespec-Launches-New-MCUX-RISC-V-Processor-That-Enables-Developers-to-Implement-Custom-Instructions-and-Add-Accelerators).

decsny replied:
That would be an acceptable Kconfig name, yes, we don't need to have mcux in the name

martinjaeger replied:
Ok. It would also match the compatible.

@Albort12138 can you please change to `CONFIG_DAC_NXP_LPDAC`? After that we should be ready to go.

decsny commented:
> @Albort12138 can you please add a comment to the binding stating the MCU series this compatible covers (so far)? 

I don't think this should be done. The IP block is agnostic to what SOC series it is used in, and therefore should not be part of the binding at all, including the description. Personally I just use grep in the dts folder when I want to quickly check what platforms a binding is used on.

martinjaeger commented:
> > @Albort12138 can you please add a comment to the binding stating the MCU series this compatible covers (so far)?
> 
> I don't think this should be done. The IP block is agnostic to what SOC series it is used in, and therefore should not be part of the binding at all, including the description. Personally I just use grep in the dts folder when I want to quickly check what platforms a binding is used on.

Ok, fine with that. Just want to have as much clarity as possible. Explaining the abbreviation would do it then.

mmahadevan108 commented on drivers/dac/dac_mcux_lpdac.c:
```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright (c) 2020 Henrik Brix Andersen <henrik@brixandersen.dk>
```
This seems like a new driver. Do we need this Copyright Line?

decsny replied:
It seems like this driver is very similar to the one that brix wrote, with only a few small differences, @Albort12138 was this intentional? If so, maybe it should say  "Based on xyz_file.c, which was: " above this line

martinjaeger replied:
> This seems like a new driver. Do we need this Copyright Line?

Yes, we need to keep this copyright line. Have you seen my comment above where I requested to add it because this driver is almost a complete copy of existing drivers written by Henrik?

Please see also: https://www.linuxfoundation.org/blog/blog/copyright-notices-in-open-source-software-projects

Adding a note "Based on xyz_file.c" is not necessary.

henrikbrixandersen replied:
Yeah, I agree with @martinjaeger.

mmahadevan108 commented on drivers/dac/dac_mcux_lpdac.c:
```diff
@@ -0,0 +1,107 @@
+
+static int mcux_lpdac_init(const struct device *dev)
+{
+	return 0;
```
Do we need a call to `DAC_Init()` from this function?

decsny replied:
DAC_Init is the only function that the SDK provides to configure the channels, so I don't think so.

danieldegrasse commented on dts/bindings/dac/nxp,lpdac.yaml:
```diff
@@ -0,0 +1,29 @@
+    type: int
+    required: true
+    description: |
+      DAC voltage reference select. The meaning of the value maybe
```
```suggestion
      DAC voltage reference select. The meaning of the value may be
```

danieldegrasse commented on dts/arm/nxp/nxp_lpc55S3x_common.dtsi:
```diff
@@ -294,6 +294,33 @@
 
+	dac0: dac@b2000 {
+		compatible = "nxp,lpdac";
+		reg = < 0xB2000 0x1000>;
```
please align casing for these reg properties with remainder of file: IE:

```suggestion
		reg = < 0xb2000 0x1000>;
```

Albort12138 committed:
boards: arm: lpcxpresso55s36: Added dac support for LPC55S36

Added dac support for the LPC55S36 board, updated lpc55xxx/soc.c to
enable clock and power for dac0.

Signed-off-by: Albort Xue <yao.xue@nxp.com>

Albort12138 committed:
samples: drivers: dac: Enable dac example on lpcxpresso55s36 board

Added overlay for lpcxpresso55s36, updated README.rst

Signed-off-by: Albort Xue <yao.xue@nxp.com>

Albort12138 committed:
tests: drivers: dac: Verify dac on lpcxpresso55s36 board

Added overlay for lpcxpresso55s36 to verify dac driver.

Signed-off-by: Albort Xue <yao.xue@nxp.com>

Albort12138 commented:
Seems I am not authorized to merge this PR, Could someone help to merge this PR?
